ekncontent/ekncontent/eknc-query-object.c

```diff
@@ -541,81 +554,86 @@ get_title_clause (EkncQueryObject *self,
+                                      corrected_terms[i], NULL);
 
       /* Discard duplicates */
       if (g_strcmp0 (term, corrected_term) == 0)
         g_clear_pointer (&corrected_term, g_free);
 
       if (term != NULL && corrected_term != NULL)
-        title_terms[i] = g_strdup_printf ("(%s%s%s)", term, XAPIAN_OP_OR, corrected_term);
+        title_terms[i] = g_strdup_printf ("(%s OR %s)", term, corrected_term);
       else if (term != NULL)
         title_terms[i] = g_steal_pointer (&term);
-      else if (corrected_term != NULL)
-        title_terms[i] = g_steal_pointer (&corrected_term);
       else
-        g_assert_not_reached ();
+        title_terms[i] = g_steal_pointer (&corrected_term);
     }
 
-  return g_strjoinv (" ", title_terms);
+  g_autofree char *parser_string = g_strjoinv (" ", title_terms);
+
+  g_autoptr(XapianQuery) retval =
+    xapian_query_parser_parse_query_full (qp, parser_string,
+                                          XAPIAN_QUERY_PARSER_FEATURE_DEFAULT |
+                                          XAPIAN_QUERY_PARSER_FEATURE_PARTIAL,
+                                          "", &error);
+  if (error != NULL)
+    {
+      g_propagate_error (error_out, error);
+      return NULL;
+    }
+
+  return g_steal_pointer (&retval);
 }
 
-static gchar *
+static XapianQuery *
 get_tags_clause (char **tags,
-                 char *join_op)
+                 XapianQueryOp join_op)
 {
   if (tags == NULL)
     return NULL;
   // Tag lists should be joined as a series of individual tag queries joined
   // by with a join operator, so an article that has any of the tags will
-  // match, e.g. [foo,bar,baz] => 'K:foo OR K:bar OR K:baz'
+  // match, e.g. [foo,bar,baz] => 'Kfoo OR Kbar OR Kbaz'
   size_t length = g_strv_length (tags);
+  if (length == 0)
+    return NULL;
+
   g_auto(GStrv) prefixed_tags = g_new0 (gchar *, length + 1);
   for (guint i = 0; i < length; i++)
-    prefixed_tags[i] = g_strdup_printf ("%s\"%s\"", XAPIAN_PREFIX_TAG, tags[i]);
-  return g_strjoinv (join_op, prefixed_tags);
+    prefixed_tags[i] = g_strdup_printf ("%s%s", XAPIAN_PREFIX_TAG, tags[i]);
+
+  return xapian_query_new_for_terms (join_op, (const char **) prefixed_tags);
 }
 
-static gchar *
+static XapianQuery *
 get_ids_clause (char **ids,
-                char *join_op)
+                XapianQueryOp join_op)
 {
   if (ids == NULL)
     return NULL;
   size_t length = g_strv_length (ids);
+  if (length == 0)
+    return NULL;
+
   g_auto(GStrv) prefixed_ids = g_new0 (gchar *, length + 1);
-  for (guint i = 0; i < length; i++)
+  for (size_t i = 0; i < length;)
     {
       const gchar *hash = eknc_utils_id_get_hash (ids[i]);
       if (hash == NULL)
         {
           g_critical ("Unexpected id structure in query object: %s", ids[i]);
-          prefixed_ids[i] = NULL;
-        }
-      else
-        {
-          prefixed_ids[i] = g_strdup_printf ("%s%s", XAPIAN_PREFIX_ID, hash);
+          continue;
```

Comment:
It looks to me like if we hit a NULL hash we'll enter an infinite loop.

I think you need a separate index for `ids` and `prefixed_ids`.